src/safeds_runner/server/_memoization_map.py

```diff
@@ -1,16 +1,266 @@
 """Module that contains the memoization logic and stats."""
+from __future__ import annotations
 
 import dataclasses
 import inspect
+import functools
+import operator
 import logging
 import sys
 import time
+import uuid
 from collections.abc import Callable
 from dataclasses import dataclass
+from multiprocessing.shared_memory import SharedMemory
 from typing import Any, TypeAlias
 
 MemoizationKey: TypeAlias = tuple[str, tuple[Any], tuple[Any]]
 
+# Contains classes that can be lazily compared, as they implement a deterministic hash
```

Comment:
This should not be needed, but I didn't find a real general solution to this.

I thought of maybe checking only the prefix of the package (allow if classes are from the `safeds` package), or checking the class names by regex?

Reply:
How do you decide which classes go in here? Do they need specific capabilities? What would happen if all classes were whitelisted?

Reply:
They just need to support deterministic hashing.
If all classes were whitelisted, they would all be stored in shared memory locations, but they would never be found by the cache, as the hash values from different processes would mismatch.
This would cause them to be added again to the cache, without the possibility of ever using them, while just consuming the memory (although they would get removed by memory pressure in the future).
This PR depends (at least partially) on https://github.com/Safe-DS/Library/pull/609 for this reason.
Classes that do not support this, could then use the fallback options.

Reply:
Could we check something like `object.__hash__(C()) == C().__hash__()`?

Reply:
That (not using id) may not be enough, as the default hash for primitive types would break this.
E.g. if the overridden hash function delegates to a hash call of an existing primitive type (or any type that does not override hash, for that matter), this again leads to a process-dependent hash value.

Reply:
For good measure, we could additionally check whether `__eq__` is overridden. If `__eq__` is structural, then `__hash__` must also be structural, since $a == b \implies hash(a) == hash(b)$.

Reply:
We would also need to disable hash randomization (`PYTHONHASHSEED`) to verify this across multiple processes.

This `__eq__` checks will fail for (un)trained models though, as they do not (and cannot, at least the classical models) override the equality.

Reply:
> This __eq__ checks will fail for (un)trained models though, as they do not (and cannot, at least the classical models) override the equality.

Yep, makes sense.

> We would also need to disable hash randomization (PYTHONHASHSEED) to verify this across multiple processes.

Would that fix the issue with the hashing of primitive values?

Reply:
> Would that fix the issue with the hashing of primitive values? Since the runner is not supposed to be run locally,

Exactly, that should fix this. We could set this env var to `0` before starting the manager / pipeline processes.

Reply:
It even seems to work for values other than 0, as long as it's fixed:

![image](https://github.com/Safe-DS/Runner/assets/2501322/e8be4cf8-ea02-4480-8925-c17a12155fcc)